apollo-ios/Design/swift-identifiable.md

```diff
@@ -0,0 +1,124 @@
+# Summary
+
+This document proposes a new GraphQL directive named `@identity`, to be used to mark a field which can uniquely identify an object.
+
+# Introduction
+
+SwiftUI makes heavy use of the [Identifiable](https://developer.apple.com/documentation/swift/identifiable) protocol, which is used to track the identity of entities across data changes. If an object with a List is replaced with a different object with the same identity, SwiftUI will animate the item changing instead of animating an insertion. Objects that do not conform to the Identifiable protocol require additional boilerplate to be usable inside SwiftUI.
+
+Apollo Client iOS could assist the developer by adding conformance to the Identifiable protocol to its generated models. Selecting the field to be used as an identity is done by adding a new GraphQL directive.
+
+A concept of identity is required to allow response objects to be cached. The various Apollo Client projects have mechanisms that allow for identifiers to be selected through additional code. The new directive could allow schema authors to assist client authors with caching.
+
+# Definition
+
+The directive is defined as:
+```graphql
+directive @identity(scope: IdentityScope = TYPE) on FIELD | FIELD_DEFINITION
+
+enum IdentityScope {
+  TYPE
+  SERVICE
+  GLOBAL
+}
+```
+
+This directive MUST be used on a field with a non-nullable scalar type. Other types are not supported.
+
+## Scope parameter
+
+The directive can have a parameter indicating the scope of the identity. They are ordered from _narrowest_ to _widest_:
+
+1. `TYPE`: The identifier is unique for the type, e.g. an auto-incrementing column from a database table.
+2. `SERVICE`: The identifier is unique across the current GraphQL Service.
+3. `GLOBAL`: The identifier is unique across all GraphQL services. This can be used by identifiers generated according to [RFC 4122](https://datatracker.ietf.org/doc/html/rfc4122) or [RFC 9562](https://datatracker.ietf.org/doc/html/rfc9562) (also known as Universally Unique IDentifiers or UUIDs).
```

Comment:
How can you guarantee that this is unique across all GraphQL services globally? Are we just assuming you won't likely ever hit a conflicting ID if you are using UUIDs? In the future of a global graph using the [`@link` directive](https://github.com/apollographql/specs/blob/main/link/v1.0/link-v1.0.md) (which is only a concept for one possible future right now), there will likely need to be a way to ensure global uniqueness, but I have a suspicion that it will end up being much more complex than this.

Reply:
There's no guarantee beyond what the RFC provides (where the chance of collision is near-zero). I've specifically worded it as "generated according to the spec" instead of only _looking_ like 128 bits.

It's still allowed to append more text to an identifier of this type (as long as the UUID portion itself is unchanged) to make the risk of collisions even smaller.